scene/2d/physics_body_2d.cpp

```diff
@@ -1111,72 +1126,149 @@ void CharacterBody2D::move_and_slide() {
+					// prevent to move against wall on ground
+					if (was_on_floor && !on_floor && result.motion.dot(up_direction) > 0 && linear_velocity.dot(up_direction) <= 0) {
 						Transform2D gt = get_global_transform();
-						if (result.motion.length() > margin) {
-							gt.elements[2] -= result.motion.slide(up_direction);
-						} else {
-							gt.elements[2] -= result.motion;
-						}
+						gt.elements[2] -= up_direction * up_direction.dot(result.motion);
 						set_global_transform(gt);
+						on_floor = true;
+						platform_rid = prev_platform_rid;
+						platform_layer = prev_platform_layer;
+
+						floor_velocity = prev_floor_velocity;
+						floor_normal = prev_floor_normal;
 						linear_velocity = Vector2();
 						return;
 					}
+					// prevent to move against wall on air
+					else if (sliding_enabled && !on_floor) {
+						motion = up_direction * up_direction.dot(result.remainder);
+						motion = motion.slide(result.collision_normal);
+					} else {
+						motion = result.remainder;
+					}
 				}
-
-				if (sliding_enabled || !on_floor) {
-					motion = result.remainder.slide(result.collision_normal);
-					linear_velocity = linear_velocity.slide(result.collision_normal);
+				// constant Speed when the slope is up
+				else if (constant_speed_on_floor && is_on_floor_only() && was_on_floor && can_apply_constant_speed && motion.dot(result.collision_normal) < 0) {
+					Vector2 motion_slide_norm = result.remainder.slide(result.collision_normal).normalized();
+					if (!motion_slide_norm.is_equal_approx(Vector2())) {
+						motion = motion_slide_norm * (motion_slide_up.length() - result.motion.slide(up_direction).length() - last_travel.slide(up_direction).length());
+					}
+				} else if ((sliding_enabled || !on_floor) && !(on_ceiling && !slide_on_ceiling && linear_velocity.dot(up_direction) > 0)) {
+					Vector2 slide_motion = result.remainder.slide(result.collision_normal);
+					if (slide_motion.dot(linear_velocity) > 0.0) {
+						motion = slide_motion;
+					} else {
+						motion = Vector2();
+					}
+					if (slide_on_ceiling && on_ceiling && linear_velocity.dot(up_direction) > 0) {
+						linear_velocity = linear_velocity.slide(result.collision_normal);
+					} else if (slide_on_ceiling && on_ceiling) { // avoid acceleration in slope when falling
+						linear_velocity = up_direction * up_direction.dot(linear_velocity);
+					}
 				} else {
 					motion = result.remainder;
+					if (on_ceiling && !slide_on_ceiling && linear_velocity.dot(up_direction) > 0) {
+						linear_velocity = linear_velocity.slide(up_direction);
+						motion = motion.slide(up_direction);
+					}
 				}
-			}
 
-			sliding_enabled = true;
+				last_travel = result.motion;
+			}
+			// apply constant speed when the slope is down
+			else if (i == 0 && constant_speed_on_floor && first_slide && was_on_floor && !Math::is_equal_approx(floor_snap_strength, 0) && _on_floor_if_snapped()) {
```

Comment:
Indeed `on_floor_is_snapped` call one extra time `move_and_collide` in test mode, I don’t think this can be avoided. The idea is the following: when you are in a downward slope and you move forward, you don’t collide because you are in the air. When `constant_speed_on_floor` is `true`, to know if I have to apply the constant speed logic, I check if the body will be on the floor if I snap it, if yes I cancel the last motion, and calculate a new motion according to the slope and the length of the velocity.

Reply:
Thanks for the explanation, I understand what you are doing now. And I missed that it was only on the first slide, so it's only done once in the worst case.

So the only case it would return false is when the character is at the border of a platform and leaving it, because we need to avoid applying constant speed in this case?

I wonder what would happen if we were just checking that snapping is enabled (the first condition in `on_floor_is_snapped`). Wouldn't that just cause a marginal difference in velocity for the one frame where you leave the platform? If I'm missing something and that would cause specific bugs or noticeable discrepancies, then I'm fine with doing the extra call to `move_and_collide`.

Reply:
> And I missed that it was only on the first slide, so it's only done once in the worst case.

Note I tried minimising the impact by adding new tests in `floor_snap` to avoid useless snapping (if up_dir = 0, if already on the floor and if the velocity is not in the ground direction)

> So the only case it would return false is when the character is at the border of a platform and leaving it, because we need to avoid applying constant speed in this case?

No, if the snapping force is not enough to maintain the body on the ground (can be desired), this will result to wrongly snap it. This will be clearer in a gif:

![without snap](https://user-images.githubusercontent.com/6397893/127741004-8b9ea20b-0242-4195-ab6b-9bbcf3711a28.gif)

Reply:
> No, if the snapping force is not enough to maintain the body on the ground (can be desired), this will result to wrongly snap it.

Ok, it makes sense to keep the check the way you implemented it then.